src/ephemeris.c

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (C) 2011-2015 Swift Navigation Inc.
+ * Contact: Fergus Noble <fergus@swift-nav.com>
+ *          Gareth McMullin <gareth@swiftnav.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#include <string.h>
+#include <libswiftnav/ephemeris.h>
+#include <libswiftnav/logging.h>
+#include <ch.h>
+
+#include "sbp.h"
+#include "track.h"
+#include "timing.h"
+#include "ephemeris.h"
+
+MUTEX_DECL(es_mutex);
+ephemeris_t es[MAX_SATS] _CCM;
+static ephemeris_t es_old[MAX_SATS] _CCM;
+static bool es_confidence[MAX_SATS] _CCM;
+
+static WORKING_AREA_CCM(wa_nav_msg_thread, 3000);
+static msg_t nav_msg_thread(void *arg)
+{
+  (void)arg;
+  chRegSetThreadName("nav msg");
+
+  while (TRUE) {
+
+    /* TODO: This should be trigged by a semaphore from the tracking loop, not
+     * just ran periodically. */
+
+
+    for (u8 i=0; i<nap_track_n_channels; i++) {
+      chThdSleepMilliseconds(100);
+      tracking_channel_t *ch = &tracking_channel[i];
+      ephemeris_t e = {.prn = ch->prn};
+
+      /* Check if there is a new nav msg subframe to process.
+       * TODO: move this into a function */
+      if ((ch->state != TRACKING_RUNNING) ||
+          (ch->nav_msg.subframe_start_index == 0))
+        continue;
+
+      /* Decode ephemeris to temporary struct */
+      __asm__("CPSID i;");
+      s8 ret = process_subframe(&ch->nav_msg, &e);
+      __asm__("CPSIE i;");
+
+      if (ret < 0) {
+        log_info("PRN %02d ret %d\n", ch->prn+1, ret);
```

Comment:
More descriptive log message here would be good - user won't know this has anything to do with ephemeris decoding at the moment.